libsolidity/experimental/analysis/TypeRegistration.cpp

```diff
@@ -0,0 +1,211 @@
+{
+	if (annotation(_typeNameExpression).typeConstructor)
+		return;
+
+	// TODO: this is not visited in the ElementaryTypeNameExpression visit - is that intentional?
+	_typeNameExpression.type().accept(*this);
+	if (auto constructor = annotation(_typeNameExpression.type()).typeConstructor)
+		annotation(_typeNameExpression).typeConstructor = constructor;
+	else
+		solAssert(m_errorReporter.hasErrors());
+}
+
+bool TypeRegistration::visit(UserDefinedTypeName const& _userDefinedTypeName)
+{
+	if (annotation(_userDefinedTypeName).typeConstructor)
+		return false;
+	auto const* declaration = _userDefinedTypeName.pathNode().annotation().referencedDeclaration;
+	if (!declaration)
+	{
+		// TODO: fatal/non-fatal
+		m_errorReporter.fatalTypeError(5096_error, _userDefinedTypeName.pathNode().location(), "Expected declaration.");
+		return false;
+	}
+	declaration->accept(*this);
+	if (!(annotation(_userDefinedTypeName).typeConstructor = annotation(*declaration).typeConstructor))
+	{
+		// TODO: fatal/non-fatal
+		m_errorReporter.fatalTypeError(9831_error, _userDefinedTypeName.pathNode().location(), "Expected type declaration.");
+		return false;
+	}
+	return true;
+}
+
+bool TypeRegistration::visit(TypeClassInstantiation const& _typeClassInstantiation)
+{
+	if (annotation(_typeClassInstantiation).typeConstructor)
+		return false;
+	_typeClassInstantiation.typeConstructor().accept(*this);
+	auto typeConstructor = annotation(_typeClassInstantiation.typeConstructor()).typeConstructor;
+	if (!typeConstructor)
+	{
+		m_errorReporter.typeError(5577_error, _typeClassInstantiation.typeConstructor().location(), "Invalid type name.");
+		return false;
+	}
+	auto* instantiations = std::visit(util::GenericVisitor{
+		[&](ASTPointer<IdentifierPath> _path) -> TypeClassInstantiations*
+		{
+			if (TypeClassDefinition const* classDefinition = dynamic_cast<TypeClassDefinition const*>(_path->annotation().referencedDeclaration))
+				return &annotation(*classDefinition).instantiations;
+			m_errorReporter.typeError(3570_error, _typeClassInstantiation.typeClass().location(), "Expected a type class.");
+			return nullptr;
+		},
+		[&](Token _token) -> TypeClassInstantiations*
+		{
+			if (auto typeClass = builtinClassFromToken(_token))
+				return &annotation().builtinClassInstantiations[*typeClass];
+			m_errorReporter.typeError(5262_error, _typeClassInstantiation.typeClass().location(), "Expected a type class.");
+			return nullptr;
+		}
+	}, _typeClassInstantiation.typeClass().name());
+
+	if (!instantiations)
+		return false;
+
+	if (
+		auto [instantiation, newlyInserted] = instantiations->emplace(*typeConstructor, &_typeClassInstantiation);
+		!newlyInserted
+	)
+	{
+		SecondarySourceLocation ssl;
+		ssl.append("Previous instantiation.", instantiation->second->location());
+		m_errorReporter.typeError(6620_error, _typeClassInstantiation.location(), ssl, "Duplicate type class instantiation.");
```

Comment:
Also, I find it odd that you can constrain type arguments of instantiations and we can distinguish instantiations with different constraints (they'll have different `Arity`), but you can't instantiate multiple times.

For example:
```solidity
pragma experimental solidity;

type Vector(Item);

class Self: Container {}

class Self: Number {}
class Self: String {}

instantiation Vector(Item: Number): Container {}
instantiation Vector(Item: String): Container {}
```
```
TypeError 6620: (171-219): Duplicate type class instantiation.
```

Is this something we're going to allow that in the future? Otherwise I don't see much point in allowing those constraints in instantiations.

I'm also generally unsure at what places we're going to allow constraints in the first place, because the prototype is pretty inconsistent about that. It disallows some things I'd expect to be possible (e.g. `type Vector(T: Number);`) while allowing some wild things (e.g. `let x: T: Number: Number: Number;`). It's incomplete so it's understandable, but it makes it hard to figure out what is supposed to be valid eventually and what not.

Reply:
This is generally impossible. How would you distinguish between the two instantiations? I.e. how would you ever be able to tell which is the right one? There can only ever be a unique instantiation for any type constructor.

In situations in which you'd want to do this, you'd, instead, move the specific logic to a new type class for the argument. I.e. have ``instantiation Vector(Item: NumberOrString): Container {}`` and appropriately define ``NumberOrString`` (we currently don't have that in the PR, but conceptually it's possible to declare ``Number`` and ``String`` subclasses of ``NumberOrString`` by implementing the interface of ``NumberOrString`` with what's available for `Number` or `String` - but in general the current PR doesn't properly handle a dependency/inheritance hierarchy of type classes yet)

Reply:
> I'm also generally unsure at what places we're going to allow constraints in the first place, because the prototype is pretty inconsistent about that. It disallows some things I'd expect to be possible (e.g. type Vector(T: Number);) while allowing some wild things (e.g. let x: T: Number: Number: Number;). It's incomplete so it's understandable, but it makes it hard to figure out what is supposed to be valid eventually and what not.

Type definitions are general and cannot be constrained to arguments of specific type classes (there's no use for doing that) It's only type class instantiations that can depend on specific argument sorts. So that part is intentional.

Stuff like ``let x:T:Number:Number:Number;`` - if that's really allowed - is just laziness so far. Remember that this is experimental and not yet meant to be perfect - it will have to change anyways, once we generalize type classes to more than one type argument.